[Lv_1] 게임 시작 구현 - 박채현

clairePark96 opened the pull request:
### 1. 신규 기능
- ```BaseballGame``` 클래스 선언
   **- enum**
```GameMessage``` : 게임에서 출력될 메시지를 클래스 내부 enum 으로 String타입의 고유값으로 지정

   **- 프로퍼티** 
```answer``` : Int 배열 상수
```numberOfAnswer``` : 답 개수

   **- 메소드**
   ```startNewSet()``` :  answer의 count가 numberOfAnswer가 될 때까지 0에서 9이내의 랜덤 Int 추출 반복하며, 동일한 수가 없을 때만 append, 게임 시작을 알리는 메시지 출력



---

### 2. 변경된 점

---

### 3. 고민했던 점
게임 메시지나 자리 수 등이 변경될 수 있다는 점, 답을 가지는 배열의 개수가 설정된 자리 수를 넘지 않도록 해야한다는 점에 고민하여 작성하였습니다.

Russell-SM-Kim committed:
Update README: 풀 리퀘스트 작성 규칙 추가

- 풀 리퀘스트 작성 규칙 추가

Russell-SM-Kim committed:
Update README : 마크다운 문법 수정

- 띄어쓰기 문제 수정

clairePark96 committed:
[Feat] 게임 시작 기능 구현

- BaseballGame 클래스 내 게임 시작 메소드 구현
- 답 배열의 count가 일정 개수가 될 때까지 랜덤숫자 추출 반복, 동일한 수가 없을 때만 append
- 게임 시작 메시지 출력

Russell-SM-Kim commented on Week2-BaseballGame/Week2-BaseballGame/baseballGame.swift:
```diff
@@ -0,0 +1,39 @@
+
+class BaseballGame {
+    
+    enum GameMessage: String {
```
우와 열거형으로 메세지 전달을 한다는건 생각도 못했어요.

근데 열거형에 케이스 이름들이 명확하게 지어지면 더 좋을 것 같아요.

예로 start -> gameStartMassage

clairePark96 replied:
지당하십니다. 수정하겠습니다!!

Russell-SM-Kim commented on Week2-BaseballGame/Week2-BaseballGame/baseballGame.swift:
```diff
@@ -0,0 +1,39 @@
+
+    
+    private var answer: [Int] = []
+    private let numberOfAnsewer = 3
```
엇 바로 오타 찾음 numberOfAnswer -> numberOfAnswer

clairePark96 replied:
상민님두 오타 numberOfAnswer -> numberOfAnswer 둘이 똑같아요ㅋ.ㅋ
감사합니다 수정하겠습니다!!

Russell-SM-Kim replied:
@chaehyunp 악 역시 자기 코드는 오타 잘 못보나봐요 ㅋㅋㅋㅋㅋ

ericKwon95 commented on Week2-BaseballGame/Week2-BaseballGame/baseballGame.swift:
```diff
@@ -0,0 +1,39 @@
+
+class BaseballGame {
+    
+    enum GameMessage: String {
```
원시값 활용 좋습니다~~
상수값은 static let으로 작성하기도 합니다
```swift
enum GameMessage {
  static let start = "Game start"
}
```

요런식으로요!

clairePark96 replied:
아하 String값이라는걸 지워야 했군요..
꼼꼼히 보질 않았네요ㅠㅠ 감사합니다!!

Crois0509 commented:
Lv1 임에도 불필요한 코드 없이 간결하고 깔끔하게 정리한 모습이 멋집니다!!
class 내부에 enum으로 메세지를 출력하는 발상도 좋은 것 같아요!!
이대로 Lv2도 힘내봐요!!

Crois0509 commented on Week2-BaseballGame/Week2-BaseballGame/baseballGame.swift:
```diff
@@ -0,0 +1,39 @@
+                answer.append(randomNumber)
+            }
+        }
+        print(GameMessage.start.rawValue)
```
enum 타입으로 지정한 메세지를 불러오셨네요!! 정말 보기 깔끔한 것 같아요ㅎㅎ
지금은 클래스 메소드에서 선언하기 때문에 GameMessage를 바로 불러올 수 있지만, 만약 클래스 밖에서 불러오면 어떻게 될까요?
`BaseballGame.GameMessage.start.rawValue`로 코드가 길어지며 가독성도, 이해력도 안 좋아지겠죠ㅠㅠ
`typealias`를 사용하면 해결할 수 있답니다!
`typealias Message = BaseballGame.GameMessage` 라고 지정하면
```swift
typealias Message = BaseballGame.GameMessage

print(Message.start.rawValue)
```
이런 식으로 짧게 표현할 수 있게 된답니다!! 타입 뿐 아니라 클로저도 저렇게 만들 수 있으니 typealias에 대해 공부해 보시면 앞으로 코드 짜실 때 좀 더 가독성이 좋은 코드를 작성하실 수 있을거에요!!

Crois0509 commented on Week2-BaseballGame/Week2-BaseballGame/baseballGame.swift:
```diff
@@ -0,0 +1,39 @@
+
+class BaseballGame {
+    
+    enum GameMessage: String {
```
기왕 enum으로 만들어 주신 김에 메소드도 하나 만들면 어떨까요?
메소드 내부에 switch로 각 case에 따른 출력값을 주고 메소드의 반환타입을 String으로 하면
GameMessage.start.rawValue 처럼 코드가 길어지지 않고 짧게 구현할 수 있을 것 같아요!
메세지 출력만을 전담하는 싱글톤 패턴을 만드는 것도 방법이겠지만, 이건 조금 과할 수 있으니 배제해도 좋겠어요

clairePark96 committed:
[Feat] 입력값 올바른지 확인후 스트라이크 / 볼 개수 카운트하여 출력

입력값이 올바르지 않은 경우 Error로 정의
입력값을 guard문으로 나누어 Error 던져주는 메소드 정의
do문에서 해당 Error 받아 올바르게 작동되는 경우 Strike/Ball 개수 파악, catch문에서 Error에 따른 오류메시지 출력

clairePark96 committed:
[Feat] 포함된 숫자 없을 경우 아웃메시지 출력

outCount 변수를 만들고 입력값 체크하는 반복문에서 포함되지 않을 때마다 +1
outCount가 입력한 자리 수와 동일할 경우 (= 모든 자리가 일치하지 않을 경우) 아웃 메시지 출력

clairePark96 committed:
[Feat] 맞출 경우에는 게임 종료, 아닐 경우 계속 진행

- while문의 조건을 게임 상태가 on일 경우, 반복해서 값을 받고 답에 대한 스트라이크와 볼 개수 출력
- 자리와 숫자가 모두 맞는 strike 변수의 값이 총 입력 값과 동일할 경우 게임 상태를 종료로 변경 -> while문 종료

clairePark96 committed:
[Test] 테스트 코드 주석 처리

스트라이크, 볼, 아웃 카운트 출력 주석 처리

clairePark96 committed:
[Refactor] enum 내 상수값을 case 대신 static let으로 수정

clairePark96 committed:
[Feat] 범위 0부터 시작, 첫 번째 자리에는 0이 올 수 없도록

- 중복 값이 아닐 경우 배열에 붙이고 첫 번째 인덱스의 값이 0일 경우 제거 구현 했었으나 연속적으로 계속 0이 나올 경우 수없이 반복
- 처음에는 0을 제외한 범위로 랜덤을 돌린 후 값이 붙은 후에는 0을 포함한 범위를 주는 방법으로 구현

clairePark96 committed:
[Feat] 게임 시작시 실행할 내용 선택

- 게임 시작 혹은 기록보기, 게임 종료 선택 가능
- 요구된 숫자 이외 다른 값 입력시 재입력 요구

clairePark96 committed:
[Rename] 스위프트 파일명 첫글자 대문자로 수정

clairePark96 committed:
[Feat] 종료 전까지 게임 계속 진행, 파일 분리

- 기록을 관리하는 다른 클래스를 만들어 게임이 계속 진행되는 동안 회차와 시도 횟수를 저장
- Model 분리

clairePark96 committed:
[Refactor] 파일 분리 및 기능 분리

- 정답 개수, 범위 등의 enum 상수 값으로 분리
- 역할이 많은 메소드 -> 기능 별도 메소드로 분리
- 기록 매니저 튜플 값에 레이블 추가
- 출력 메소드 복잡, 각 enum에서 메시지 출력 메소드로 인해 기능 과다 -> 출력 매니저 구조체 선언
